### PR TITLE
Adding basic benchmarks for IntObjectHashMap

### PR DESCRIPTION
**Pull request by nmittler**
Motivation:

It needs to be fast :)

Modifications:

Added a simple benchmark to the microbench module.

Result:

Yay, benchmarks!

### CODE REVIEW

**Comment by nmittler** [comment]
@normanmaurer @Scottmitch @buchgr this is what I used to generate the benchmarks for #3613

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Benchmark
+    public void agronaPut(Blackhole bh) {
+        for (int key : keys) {
+            bh.consume(agronaMap.put(key, VALUE));
+        }
+    }
+
+    public static void main(String[] args) throws RunnerException {
```
Comment (Scottmitch) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Can you use the runner/options provided by `AbstractMicrobenchmark`?  Should allow you to run as JUNIT test out of the box...you can configure most of these with annotations...or override the `jvmArgs()` if not.

File: microbench/pom.xml [hunk 2]
```diff
@@ -73,14 +73,19 @@
     <dependency>
       <groupId>org.openjdk.jmh</groupId>
       <artifactId>jmh-core</artifactId>
-      <version>0.9</version>
+      <version>1.7.1</version>
```
Comment (Scottmitch) [inline microbench/pom.xml]: Been a while since we updated I guess...do your features depend upon this update though? I'm not against is...just curious.

Comment (nmittler) [inline microbench/pom.xml]: No this wasn't required ... just an artifact of having some initial trouble getting the benchmarks to run.  Probably not a bad idea to upgrade though.

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java (hunk 1 above)
Comment (nmittler) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Done.

File: microbench/pom.xml (hunk 2 above)
Comment (normanmaurer) [inline microbench/pom.xml]: +1 :+1:

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.Set;
+
+public class IntObjectHashMapBenchmark extends AbstractMicrobenchmark {
+    private static final Long VALUE = Long.MAX_VALUE;
+
+    @Param({ "10", "100", "1000", "10000", "100000" })
+    int size;
+
+    int[] keys;
+    IntObjectHashMap<Long> nettyMap = new IntObjectHashMap<Long>();
```
Comment (normanmaurer) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: why are these package private and not just private ?

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void agronaLookup(Blackhole bh) {
+        for (int key : keys) {
+            bh.consume(agronaMap.get(key));
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
```
Comment (normanmaurer) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Remove ?

**Comment by trustin** [comment]
LGTM - Please address @normanmaurer and then feel free to cherry-pick.

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java (hunk 3 above)
Comment (nmittler) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: They should be private ... done.

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java (hunk 4 above)
Comment (nmittler) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: It's useful for running within an IDE.  I was actually wondering why the other benchmarks don't have this?

Comment (normanmaurer) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Honestly doing benchmarks with an idea is kind of pointless if you ask me.

Comment (nmittler) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Done.

**Comment by nmittler** [comment]
@normanmaurer @trustin @Scottmitch Are we ok having these benchmarks compare our map against Agrona's or should I get rid of the Agrona dependency altogether?

**Comment by nmittler** [comment]
PTAL, I've addressed comments and added several new benchmarks.

File: microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+    @Param
+    public KeyDistribution keyDistribution;
+
+    private Environment environment;
+
+    @Setup(Level.Trial)
+    public void setup() {
+        switch(mapType) {
+            case AGRONA: {
```
Comment (Scottmitch) [inline microbench/src/test/java/io/netty/microbenchmark/common/IntObjectHashMapBenchmark.java]: Ooops I just submitted a PR and told @louiscryan that we haven't been indenting these...but now I look at is seems like we have been (for the majority of cases).  Sorry @louiscryan .. first case I looked at was not indented :/

**Comment by Scottmitch** [comment]
@nmittler - Looks good to me.  I'm fine with having external dependencies in the micro benchmark.  Makes sense to be able to compare what we have vs what others have to see if we can improve or adopt better stuff.

**Comment by nmittler** [comment]
cherry-picked into master and 4.1

**Pull request closed without merging**